### ISSUE TITLE
Can't $Expand on collection property 

### ISSUE THREAD

**Issue by a-elnajjar** [OData/WebApi#1069]
I have these Entities Client and Assessments  . when I use string Query to Expand client to include Assessment I have this error 
`url 'http://localhost:20961/Clients?$expand=Assessments'`

> {
>   "error": {
>     "code": "",
>     "message": "An error has occurred.",
>     "innererror": {
>       "message": "Argument types do not match",
>       "type": "System.ArgumentException",
>       "stacktrace": "   at System.Linq.Expressions.Expression.Condition(Expression test, Expression ifTrue, Expression ifFalse)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitConditional(ConditionalExpression node)\r\n   at System.Linq.Expressions.ConditionalExpression.Accept(ExpressionVisitor visitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)\r\n   at Remotion.Linq.Clauses.FromClauseBase.TransformExpressions(Func`2 transformation)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ExpressionTransformingQueryModelVisitor`1.VisitMainFromClause(MainFromClause fromClause, QueryModel queryModel)\r\n   at Remotion.Linq.Clauses.MainFromClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)\r\n   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.Rewrite(QueryModel queryModel, QueryModel parentQueryModel)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitSubQuery(SubQueryExpression expression)\r\n   at Remotion.Linq.Clauses.Expressions.SubQueryExpression.Accept(ExpressionVisitor visitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitConditional(ConditionalExpression node)\r\n   at System.Linq.Expressions.ConditionalExpression.Accept(ExpressionVisitor visitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitMemberAssignment(MemberAssignment node)\r\n   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)\r\n   at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitMemberAssignment(MemberAssignment node)\r\n   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)\r\n   at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)\r\n   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)\r\n   at Remotion.Linq.Clauses.SelectClause.TransformExpressions(Func`2 transformation)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ExpressionTransformingQueryModelVisitor`1.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.NavigationRewritingQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)\r\n   at Remotion.Linq.Clauses.SelectClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)\r\n   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)\r\n   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.Rewrite(QueryModel queryModel, QueryModel parentQueryModel)\r\n   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel)\r\n   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)\r\n   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, INodeTypeProvider nodeTypeProvider, IDatabase database, ILogger logger, Type contextType)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass19_0`1.<CompileQuery>b__0()\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQuery[TResult](Expression query)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)\r\n   at Remotion.Linq.QueryableBase`1.System.Collections.IEnumerable.GetEnumerator()\r\n   at System.Web.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteResourceSet(IEnumerable enumerable, IEdmTypeReference resourceSetType, ODataWriter writer, ODataSerializerContext writeContext)\r\n   at System.Web.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext)\r\n   at System.Web.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteObject(Object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)\r\n   at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStream(Type type, Object value, Stream writeStream, HttpContent content, HttpContentHeaders contentHeaders)\r\n   at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStreamAsync(Type type, Object value, Stream writeStream, HttpContent content, TransportContext transportContext, CancellationToken cancellationToken)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Net.Http.HttpContent.<LoadIntoBufferAsyncCore>d__49.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Owin.HttpMessageHandlerAdapter.<BufferResponseContentAsync>d__13.MoveNext()"
>     }
>   }
> }

but when I do this URL `'http://localhost:20961/Assessments?$expand=Client'` it works. 
Client class
```

   public class Client : BaseEntity
    {
        public Client()
        {
            Assessments = new HashSet<Assessment>();
            AssessmentScoreHistory = new HashSet<AssessmentScoreHistory>();
            ClientTags = new HashSet<ClientTag>();
            Notifications = new HashSet<Notification>();
            ReportNotifications = new HashSet<ReportNotification>();
            ReportHistory = new HashSet<ReportHistory>();
            CreatedDate = DateTime.UtcNow;
        }
        public Guid Id { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Email { get; set; }
        public string ParticipantId { get; set; }
        public DateTime? BirthDate { get; set; }
        public int Gender { get; set; }
        public byte? Age { get; set; }
        public int Status { get; set; }
        public string Notes { get; set; }
        public virtual Language DefaultLanguage { get; set; }
        public virtual User User { get; set; }
        public virtual ICollection<Tag> Tags { get; set; }
        public virtual ICollection<Notification> Notifications { get; set; }
        public virtual ICollection<ReportNotification> ReportNotifications { get; set; }
        public virtual ICollection<Assessment> Assessments { get; set; }
        public virtual ICollection<AssessmentScoreHistory> AssessmentScoreHistory { get; set; }
        public virtual ICollection<ReportHistory> ReportHistory { get; set; }
        public HashSet<ClientTag> ClientTags { get; set; }
}

    public class Assessment : BaseEntity
    {

        public Assessment(){}
        public Guid Id { get; set; }
        public int? Status { get; set; }
        public int AssessmentNumber { get; set; }
        public virtual Language Language { get; set; }
        public virtual AssessmentType AssessmentType { get; set; }
        public User User { get; set; }
        public Guid? ClientId { get; set; }
        public virtual Client Client { get; set; }
        public String ClientName { get; set; }
        public DateTime? AssessmentDate { get; set; }
        public virtual Catalog Catalog { get; set; }
        public string CatalogId { get; set; }
        public virtual Rater Rater { get; set; }
        public virtual AssessmentData AssessmentData { get; set; }
        public virtual AssessmentResume AssessmentResume { get; set; }
        public virtual AssessmentScoreHistory AssessmentScoreHistory { get; set; }
        public string JSONData { get; set; }
        //Ignore in Database
        public Client GetClient { get; set; }
}
```
from DbContaxt
           //client
            modelBuilder.Entity<Client>().ToTable("Clients");
            modelBuilder.Entity<Client>().HasKey(x => x.Id);
            modelBuilder.Entity<Client>().Property(x => x.Id).ForSqlServerHasDefaultValueSql("newid()"); //<-- this line should create new Guid in insert
            modelBuilder.Entity<Client>().Ignore(x => x.DisplayName);
            modelBuilder.Entity<Client>().Ignore(x => x.DisplayAgeInfo);
            modelBuilder.Entity<Client>().Property(x => x.FirstName).HasMaxLength(50);
            modelBuilder.Entity<Client>().Property(x => x.LastName).HasMaxLength(50);
            modelBuilder.Entity<Client>().Property(x => x.ParticipantId).HasMaxLength(50);
            modelBuilder.Entity<Client>().Property(x => x.Email).HasMaxLength(100);
            modelBuilder.Entity<Client>().HasOne(x => x.DefaultLanguage).WithMany(x => x.Clients).HasForeignKey("DefaultLanguageId");
            modelBuilder.Entity<Client>().HasOne(x => x.User).WithMany(x => x.Clients).HasForeignKey("UserId");
            modelBuilder.Entity<Client>().HasMany(c => c.ClientTags).WithOne(c => c.Client).HasForeignKey("ClientId");
            modelBuilder.Entity<Client>().HasMany(c => c.Assessments).WithOne(c => c.Client).HasForeignKey("ClientId"); 
      //2-Assessment
            modelBuilder.Entity<Assessment>().ToTable("Assessments");
            modelBuilder.Entity<Assessment>().HasKey(x => x.Id);
            modelBuilder.Entity<Assessment>().Property(x => x.Id).ForSqlServerHasDefaultValueSql("newid()");
            modelBuilder.Entity<Assessment>().Property(x => x.AssessmentNumber).ValueGeneratedOnAdd();
            modelBuilder.Entity<Assessment>().Ignore(x => x.GetClient);
            modelBuilder.Entity<Assessment>().Property(x => x.JSONData);
            modelBuilder.Entity<Assessment>().Property(x => x.AssessmentDate).HasColumnType("datetime");
            modelBuilder.Entity<Assessment>().HasOne(x => x.Language).WithMany(x => x.Assessments).HasForeignKey("LanguageId");
            modelBuilder.Entity<Assessment>().HasOne(x => x.AssessmentData).WithOne(x => x.Assessment).HasForeignKey<Assessment>(x => x.Id);
            modelBuilder.Entity<Assessment>().HasOne(x => x.AssessmentResume).WithOne(x => x.Assessment).HasForeignKey<Assessment>(x => x.Id);
            modelBuilder.Entity<Assessment>().HasOne(x => x.AssessmentScoreHistory).WithOne(x => x.Assessment).HasForeignKey<Assessment>(x => x.Id);
            modelBuilder.Entity<Assessment>().HasOne(x => x.Catalog).WithMany(x => x.Assessments).HasForeignKey("CatalogId");
            modelBuilder.Entity<Assessment>().HasOne(x => x.User).WithMany(x => x.Assessments).HasForeignKey("UserId");
            modelBuilder.Entity<Assessment>().HasOne(x => x.Client).WithMany(x => x.Assessments).HasForeignKey("ClientId");
            ApplyJsonPropertyConfig<Assessment>(modelBuilder);
```
we are using 
Entity framework core1.2.1, 
Microsoft.AspNet.OData 6.0.0
Microsoft.OData.Core  7.3.1
Microsoft.OData.Edm  7.3.1

**Comment by a-elnajjar** [comment]
public class Client : BaseEntity
    {
        public Client()
        {
            Assessments = new HashSet<Assessment>();
            AssessmentScoreHistory = new HashSet<AssessmentScoreHistory>();
            ClientTags = new HashSet<ClientTag>();
            Notifications = new HashSet<Notification>();
            ReportNotifications = new HashSet<ReportNotification>();
            ReportHistory = new HashSet<ReportHistory>();
            CreatedDate = DateTime.UtcNow;
        }
        public Guid Id { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Email { get; set; }
        public string ParticipantId { get; set; }
        public DateTime? BirthDate { get; set; }
        public int Gender { get; set; }
        public byte? Age { get; set; }
        public int Status { get; set; }
        public string Notes { get; set; }
        public virtual Language DefaultLanguage { get; set; }
        public virtual User User { get; set; }
        public virtual ICollection<Tag> Tags { get; set; }
        public virtual ICollection<Notification> Notifications { get; set; }
        public virtual ICollection<ReportNotification> ReportNotifications { get; set; }
        public virtual ICollection<Assessment> Assessments { get; set; }
        public virtual ICollection<AssessmentScoreHistory> AssessmentScoreHistory { get; set; }
        public virtual ICollection<ReportHistory> ReportHistory { get; set; }
        public HashSet<ClientTag> ClientTags { get; set; }
}

    public class Assessment : BaseEntity
    {

        public Assessment(){}
        public Guid Id { get; set; }
        public int? Status { get; set; }
        public int AssessmentNumber { get; set; }
        public virtual Language Language { get; set; }
        public virtual AssessmentType AssessmentType { get; set; }
        public User User { get; set; }
        public Guid? ClientId { get; set; }
        public virtual Client Client { get; set; }
        public String ClientName { get; set; }
        public DateTime? AssessmentDate { get; set; }
        public virtual Catalog Catalog { get; set; }
        public string CatalogId { get; set; }
        public virtual Rater Rater { get; set; }
        public virtual AssessmentData AssessmentData { get; set; }
        public virtual AssessmentResume AssessmentResume { get; set; }
        public virtual AssessmentScoreHistory AssessmentScoreHistory { get; set; }
        public string JSONData { get; set; }
        //Ignore in Database
        public Client GetClient { get; set; }
}
from DbContaxt
//client
modelBuilder.Entity().ToTable("Clients");
modelBuilder.Entity().HasKey(x => x.Id);
modelBuilder.Entity().Property(x => x.Id).ForSqlServerHasDefaultValueSql("newid()"); //<-- this line should create new Guid in insert
modelBuilder.Entity().Ignore(x => x.DisplayName);
modelBuilder.Entity().Ignore(x => x.DisplayAgeInfo);
modelBuilder.Entity().Property(x => x.FirstName).HasMaxLength(50);
modelBuilder.Entity().Property(x => x.LastName).HasMaxLength(50);
modelBuilder.Entity().Property(x => x.ParticipantId).HasMaxLength(50);
modelBuilder.Entity().Property(x => x.Email).HasMaxLength(100);
modelBuilder.Entity().HasOne(x => x.DefaultLanguage).WithMany(x => x.Clients).HasForeignKey("DefaultLanguageId");
modelBuilder.Entity().HasOne(x => x.User).WithMany(x => x.Clients).HasForeignKey("UserId");
modelBuilder.Entity().HasMany(c => c.ClientTags).WithOne(c => c.Client).HasForeignKey("ClientId");
modelBuilder.Entity().HasMany(c => c.Assessments).WithOne(c => c.Client).HasForeignKey("ClientId");
//2-Assessment
modelBuilder.Entity().ToTable("Assessments");
modelBuilder.Entity().HasKey(x => x.Id);
modelBuilder.Entity().Property(x => x.Id).ForSqlServerHasDefaultValueSql("newid()");
modelBuilder.Entity().Property(x => x.AssessmentNumber).ValueGeneratedOnAdd();
modelBuilder.Entity().Ignore(x => x.GetClient);
modelBuilder.Entity().Property(x => x.JSONData);
modelBuilder.Entity().Property(x => x.AssessmentDate).HasColumnType("datetime");
modelBuilder.Entity().HasOne(x => x.Language).WithMany(x => x.Assessments).HasForeignKey("LanguageId");
modelBuilder.Entity().HasOne(x => x.AssessmentData).WithOne(x => x.Assessment).HasForeignKey(x => x.Id);
modelBuilder.Entity().HasOne(x => x.AssessmentResume).WithOne(x => x.Assessment).HasForeignKey(x => x.Id);
modelBuilder.Entity().HasOne(x => x.AssessmentScoreHistory).WithOne(x => x.Assessment).HasForeignKey(x => x.Id);
modelBuilder.Entity().HasOne(x => x.Catalog).WithMany(x => x.Assessments).HasForeignKey("CatalogId");
modelBuilder.Entity().HasOne(x => x.User).WithMany(x => x.Assessments).HasForeignKey("UserId");
modelBuilder.Entity().HasOne(x => x.Client).WithMany(x => x.Assessments).HasForeignKey("ClientId");
ApplyJsonPropertyConfig(modelBuilder);

Assessments DML
	[Id] [uniqueidentifier] NOT NULL,
	[Status] [int] NULL,
	[AssessmentNumber] [int] IDENTITY(1,1) NOT NULL,
	[ClientId] [uniqueidentifier] NULL,
	[ClientName] [nvarchar](max) NULL,
	[AssessmentDate] [datetime] NULL,
	[JSONData] [nvarchar](max) NULL,
	[CreatedDate] [datetime] NOT NULL,
	[UpdatedDate] [datetime] NULL,
	[DeletedDate] [datetime] NULL,
	[IsDeleted] [bit] NOT NULL,
	[AssessmentTypeId] [uniqueidentifier] NULL,
	[CatalogId] [nvarchar](50) NULL,
	[LanguageId] [int] NULL,
	[UserId] [uniqueidentifier] NULL,
 CONSTRAINT [PK_dbo.Assessments] PRIMARY KEY CLUSTERED 
Clients DML
	[Id] [uniqueidentifier] NOT NULL,
	[FirstName] [nvarchar](50) NULL,
	[LastName] [nvarchar](50) NULL,
	[Email] [nvarchar](100) NULL,
	[ParticipantId] [nvarchar](50) NULL,
	[BirthDate] [datetime] NULL,
	[Gender] [int] NOT NULL,
	[Age] [tinyint] NULL,
	[Status] [int] NOT NULL,
	[Notes] [nvarchar](max) NULL,
	[CreatedDate] [datetime] NOT NULL,
	[UpdatedDate] [datetime] NULL,
	[DeletedDate] [datetime] NULL,
	[IsDeleted] [bit] NOT NULL,
	[DefaultLanguageId] [int] NULL,
	[UserId] [uniqueidentifier] NOT NULL,
 CONSTRAINT [PK_dbo.Clients] PRIMARY KEY CLUSTERED 

we are using 
Entity framework core1.2.1, 
Microsoft.AspNet.OData 6.0.0
Microsoft.OData.Core  7.3.1
Microsoft.OData.Edm  7.3.1

**Comment by xiaoxin1983** [comment]
+1

**Comment by AlanWong-MS** [comment]
@a-elnajjar, can you verify if this issue has been resolved with PR #1081? Thanks.
